trieste/acquisition/function.py

```diff
@@ -509,6 +512,156 @@ def prepare_acquisition_function(
+        :exc:`~tf.errors.InvalidArgumentError` if used with a batch size greater than one.
+    """
+
+    def acquisition(x: TensorType) -> TensorType:
+        tf.debugging.assert_shapes(
+            [(x, [..., 1, None])],
+            message="This acquisition function only supports batch sizes of one.",
+        )
+        normal = tfp.distributions.Normal(
+            loc=tf.zeros(shape=1, dtype=x.dtype), scale=tf.ones(shape=1, dtype=x.dtype)
+        )
+
+        def Psi(a, b, mean, std) -> TensorType:
+            """
+            Generic Expected Improvement defined at Eq. 19 in :cite:`yang2019efficient`,
+            calculate the expected improvement on reference a, integrating from b
+            :param a: shape [num_cells, out_dim] expected improvement reference point
+            :param b: shape [num_cells, out_dim] integral upper bounds
+            :param mean: shape [..., out_dim]
+            :param std: shape [..., out_dim]
+            """
+            return std * normal.prob((b - mean) / std) + (mean - a) * (
+                1 - normal.cdf((b - mean) / std)
+            )
+
+        def nu(lb, ub, mean, std) -> TensorType:
+            """
+            A special expected improvement given fixed improvement,
+            defined at Eq. 25 in :cite:`yang2019efficient`
+            :param lb: shape [..., num_cells, out_dim] slice lower bound
+            :param ub: shape [..., num_cells, out_dim] slice upper bound
+            :param mean: shape [..., out_dim]
+            :param std: shape [..., out_dim]
+            """
+            return (ub - lb) * (1 - normal.cdf((ub - mean) / std))
+
+        candidate_mean, candidate_var = model.predict(tf.squeeze(x, -2))
+        candidate_std = tf.sqrt(candidate_var)
+
+        # calc ehvi assuming maximization
+        neg_candidate_mean = -tf.expand_dims(candidate_mean, 1)  # [..., 1, out_dim]
+        candidate_std = tf.expand_dims(candidate_std, 1)  # [..., 1, out_dim]
+
+        lb_points, ub_points = pareto.hypercell_bounds(
+            tf.constant([-inf] * candidate_mean.shape[-1], dtype=x.dtype), reference_point
+        )
+
+        neg_lb_points, neg_ub_points = -ub_points, -lb_points
+
+        neg_ub_points = tf.minimum(
+            neg_ub_points, 1e10
+        )  # this maximum: 1e10 is heuristically chosen
+
+        psi_ub = Psi(
+            neg_lb_points, neg_ub_points, neg_candidate_mean, candidate_std
+        )  # [..., num_cells, out_dim]
+        psi_lb = Psi(
+            neg_lb_points, neg_lb_points, neg_candidate_mean, candidate_std
+        )  # [..., num_cells, out_dim]
+
+        psi_lb2ub = tf.maximum(psi_lb - psi_ub, 0.0)  # [..., num_cells, out_dim]
+        nu_contrib = nu(neg_lb_points, neg_ub_points, neg_candidate_mean, candidate_std)
+
+        # get stacked factors of Eq. 45
+        # [2^m, indices_at_dim]
+        cross_index = tf.constant(
+            list(product(*[[0, 1] for _ in range(reference_point.shape[-1])]))
```

Comment:
It is actually possible to generate this directly in tf via:

```
dim = reference_point.shape[-1]
cross_index = tf.reshape(tf.stack(tf.meshgrid(*[[0,1]]*dim), axis=-1), (-1,dim))
```

but your way is probably clearer. Still, you can simplify it further to:

```suggestion
            list(product(*[[0, 1]] * reference_point.shape[-1]))
```